File: server/utils/network.go
package utils

import (
	"encoding/json"
	"os"
	"os/exec"
	"regexp"
	"strings"
)

const ResolvConfBackup = "/etc/resolv.conf.bak"

func GetDefaultGateway() (string, error) {
	cmd := exec.Command("ip", "r")

	// Capture standard output and standard error
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", err
	}

	// get line starting with default via and get the following ip in that line
	return strings.Split(strings.Split(string(output), "default via ")[1], " ")[0], nil
}

func BackupResolvConf() {
	if !FileExists(ResolvConfBackup) {
		// copy exising resolv.conf to resolv.conf.bak
		_, err := RunCommand(false, "/bin/cp", "/etc/resolv.conf", ResolvConfBackup)
		if !FileExists(ResolvConfBackup) {
			LogError("Error creating "+ResolvConfBackup, err)
		}
	}
}

func RestoreResolvConf() {
	// copy exising resolv.conf.bak to resolv.conf. don't use cp, read content from resolv.conf.bak
	if _, err := os.Stat(ResolvConfBackup); !os.IsNotExist(err) {
		fileContent, _ := os.ReadFile(ResolvConfBackup)
		if err := os.WriteFile("/etc/resolv.conf", fileContent, 0644); err != nil {
			LogError("Error updating /etc/resolv.conf", err)
		}
	}
}

func GetHostGateway() string {
	out, _ := RunCommand(false, "/sbin/ip", "route", "show", "default")

	// get line starting with default via and get the following ip in that line
	gw := strings.Split(strings.Split(string(out), "default via ")[1], " ")[0]
	if gw != "" {
		return gw
	}

	if _, err := os.Stat(ResolvConfBackup); !os.IsNotExist(err) {
		fileContent, err := os.ReadFile(ResolvConfBackup)
		if err == nil {
			// extract first nameserver as host gateway
			lines := strings.Split(string(fileContent), "\n")
			for _, line := range lines {
				if strings.HasPrefix(line, "nameserver") {
					return strings.Split(line, " ")[1]
				}
			}
		}
	}

	return ""
}

func GetIpInfo(ipInfo map[string]interface{}) error {
	LogLn("get ip info")
	// https://worldtimeapi.org/api/ip
	cmd := exec.Command("/usr/bin/wget", "-q", "-O", "-", "https://ipinfo.io/json")
	out, err := cmd.CombinedOutput()
	if err != nil {
		LogLn(string(out))
		LogError("Error getting ip info", err)
		return err
	}

	for k := range ipInfo {
		delete(ipInfo, k)
	}

	err = json.Unmarshal(out, &ipInfo)
	if err != nil {
		return err
	}

	return nil
}

func GetIpV4Addr(dev string, stripMask bool) string {
	cmd := exec.Command("/sbin/ip", "a", "s", dev)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return ""
	}
	// regex for ip address
	regex, _ := regexp.Compile(`([0-9]{1,3}(\.[0-9]{1,3}){3})\/[0-9]{1,2}`)

	// use regex to get ip address
	addr := regex.FindString(string(out))

	if stripMask {
		addr = strings.Split(addr, "/")[0]
	}
	return addr
}

File: server/utils/network.py
import os
import subprocess
import json
import re
from .exec import run_command
from .file import file_exists
from .log import log_ln, log_error

RESOLV_CONF_BACKUP = "/etc/resolv.conf.bak"

def get_default_gateway():
    try:
        output = subprocess.check_output(["ip", "r"], text=True)
        for line in output.split("\n"):
            if line.startswith("default via"):
                return line.split(" ")[2], None
    except Exception as e:
        return "", e
    return "", None

def backup_resolv_conf():
    if not file_exists(RESOLV_CONF_BACKUP):
        run_command(False, "/bin/cp", "/etc/resolv.conf", RESOLV_CONF_BACKUP)
        if not file_exists(RESOLV_CONF_BACKUP):
            log_error(f"Error creating {RESOLV_CONF_BACKUP}", Exception("cp failed"))

def restore_resolv_conf():
    if file_exists(RESOLV_CONF_BACKUP):
        with open(RESOLV_CONF_BACKUP, "r") as f:
            content = f.read()
        with open("/etc/resolv.conf", "w") as f:
            f.write(content)

def get_host_gateway():
    out, _ = run_command(False, "/sbin/ip", "route", "show", "default")
    if out:
        for line in out.split("\n"):
            if line.startswith("default via"):
                gw = line.split(" ")[2]
                if gw:
                    return gw
    
    if file_exists(RESOLV_CONF_BACKUP):
        with open(RESOLV_CONF_BACKUP, "r") as f:
            for line in f:
                if line.startswith("nameserver"):
                    return line.split(" ")[1].strip()
    return ""

def get_ip_info(ip_info_dict):
    log_ln("get ip info")
    try:
        import urllib.request
        import ssl
        ctx = ssl.create_default_context()
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE
        req = urllib.request.Request("https://ipinfo.io/json")
        with urllib.request.urlopen(req, context=ctx, timeout=10) as response:
            out = response.read().decode('utf-8')
        ip_info_dict.clear()
        ip_info_dict.update(json.loads(out))
        return None
    except Exception as e:
        log_error("Error getting ip info", e)
        return e

def get_ipv4_addr(dev, strip_mask=False):
    try:
        out = subprocess.check_output(["/sbin/ip", "a", "s", dev], text=True)
        match = re.search(r'([0-9]{1,3}(?:\.[0-9]{1,3}){3})\/[0-9]{1,2}', out)
        if match:
            addr = match.group(0)
            if strip_mask:
                addr = addr.split("/")[0]
            return addr
    except Exception:
        pass
    return ""
